src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs

```diff
@@ -51,6 +51,309 @@ public class C
+                //     public int Property { get; init; }
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion8, "init").WithArguments("init-only setters", "9.0").WithLocation(4, 32),
+                // (8,22): error CS0617: 'Property2' is not a valid named attribute argument. Named attribute arguments must be fields which are not readonly, static, or const, or read-write properties which are public and not static.
+                // [Test(Property = 42, Property2 = 43)]
+                Diagnostic(ErrorCode.ERR_BadNamedAttributeArgument, "Property2").WithArguments("Property2").WithLocation(8, 22)
+                );
+        }
+
+        [Fact, WorkItem(50245, "https://github.com/dotnet/roslyn/issues/50245")]
+        public void TestCSharp8_ConsumptionFromCompilationReference()
+        {
+            string lib_cs = @"
+public class C
+{
+    public string Property { get; init; }
+    public string Property2 { get; }
+}
+";
+            var lib = CreateCompilation(new[] { lib_cs, IsExternalInitTypeDefinition }, parseOptions: TestOptions.Regular8, assemblyName: "lib");
+
+            string source = @"
+public class D
+{
+    void M()
+    {
+        _ = new C() { Property = string.Empty, Property2 = string.Empty };
+    }
+}
+";
+
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8, references: new[] { lib.ToMetadataReference() }, assemblyName: "comp");
+            comp.VerifyEmitDiagnostics(
+                // (6,23): error CS8400: Feature 'init-only setters' is not available in C# 8.0. Please use language version 9.0 or greater.
+                //         _ = new C() { Property = string.Empty, Property2 = string.Empty };
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion8, "Property").WithArguments("init-only setters", "9.0").WithLocation(6, 23),
+                // (6,48): error CS0200: Property or indexer 'C.Property2' cannot be assigned to -- it is read only
+                //         _ = new C() { Property = string.Empty, Property2 = string.Empty };
+                Diagnostic(ErrorCode.ERR_AssgReadonlyProp, "Property2").WithArguments("C.Property2").WithLocation(6, 48)
+                );
+        }
+
+        [Theory, CombinatorialData, WorkItem(50245, "https://github.com/dotnet/roslyn/issues/50245")]
+        public void TestCSharp8_ConsumptionWithDynamicArgument(bool useMetadataImage)
+        {
+            string lib_cs = @"
+public class C
+{
+    public string Property { get; init; }
+    public string Property2 { get; }
+    public C(int i) { }
+}
+";
+            var lib = CreateCompilation(new[] { lib_cs, IsExternalInitTypeDefinition }, parseOptions: TestOptions.Regular9);
+
+            string source = @"
+public class D
+{
+    void M(dynamic d)
+    {
+        _ = new C(d) { Property = string.Empty, Property2 = string.Empty };
```

Comment:
>string.Empty [](start = 34, length = 12)

Consider also a test where this is dynamic.

Reply:
The reason I added this test with dynamic argument to constructor is because `BindClassCreationExpression` has special handling for this scenario.